src/protocol/libp2p/kademlia/mod.rs

```diff
@@ -291,13 +293,26 @@ impl Kademlia {
         tracing::trace!(target: LOG_TARGET, ?peer, ?query, "disconnect peer");
 
         if let Some(query) = query {
-            self.engine.register_response_failure(query, peer);
+            self.engine.register_peer_failure(query, peer);
         }
 
+        // Apart from the failing query, we need to fail all other pending queries for the peer
+        // being disconnected.
         if let Some(PeerContext { pending_actions }) = self.peers.remove(&peer) {
             pending_actions.into_iter().for_each(|(_, action)| {
-                if let PeerAction::SendFindNode(query_id) = action {
-                    self.engine.register_response_failure(query_id, peer);
+                let query_id = match action {
+                    // `SendFindNode` includes `FIND_NODE`, `GET_VALUE` and `GET_PROVIDERS`
+                    // queries.
+                    PeerAction::SendFindNode(query_id)
+                    | PeerAction::SendPutValue(query_id, _)
+                    | PeerAction::SendAddProvider(query_id, _) => query_id,
+                };
+
+                // Don't report failure twice for the same `query_id` if it was already reported
+                // above. (We can still have other pending queries for the peer that
+                // need to be reported.)
+                if Some(query_id) != query {
```

Comment:
dq: Doesn't the `if Some(query_id)` shadows the previous `let query_id = peeraction.query_id()`?